ffmpeg_new.c

```diff
@@ -0,0 +1,555 @@
+	int64_t channel_layout = 0;
+	AVCodec *codec;
+	AVCodecContext *cc = NULL;
+	AVFormatContext *ic = NULL;
+	SwrContext *swr = NULL;
+
+	ffmpeg_init();
+
+	err = avformat_open_input(&ic, ip_data->filename, NULL, NULL);
+	if (err < 0) {
+		d_print("av_open failed: %d\n", err);
+		return -IP_ERROR_FILE_FORMAT;
+	}
+
+	do {
+		err = avformat_find_stream_info(ic, NULL);
+		if (err < 0) {
+			d_print("unable to find stream info: %d\n", err);
+			err = -IP_ERROR_FILE_FORMAT;
+			break;
+		}
+
+		err = av_find_best_stream(ic, AVMEDIA_TYPE_AUDIO, -1, -1, &codec, 0);
+		if (err < 0) {
+			if (err == AVERROR_STREAM_NOT_FOUND) {
+				d_print("could not find audio stream\n");
+				err = -IP_ERROR_FILE_FORMAT;
+			} else if (err == AVERROR_DECODER_NOT_FOUND) {
+				d_print("codec not found\n");
+				err = -IP_ERROR_UNSUPPORTED_FILE_TYPE;
+			} else {
+				err = -IP_ERROR_INTERNAL;
+			}
+			break;
+		}
+		stream_index = err;
+
+		cc = avcodec_alloc_context3(codec);
+		if (!cc) {
+			d_print("could not allocate decodec context\n");
+			err = -IP_ERROR_INTERNAL;
+			break;
+		}
+		if (ffmpeg_initialize_context(cc, ic->streams[stream_index]) < 0) {
+			d_print("could not initialize decodec context\n");
+			err = -IP_ERROR_INTERNAL;
+			break;
+		}
+
+		if (codec->capabilities & CODEC_CAP_TRUNCATED)
+			cc->flags |= CODEC_FLAG_TRUNCATED;
+
+		if (avcodec_open2(cc, codec, NULL) < 0) {
+			d_print("could not open codec: %d, %s\n", cc->codec_id,
+					avcodec_get_name(cc->codec_id));
+			err = -IP_ERROR_UNSUPPORTED_FILE_TYPE;
+			break;
+		}
+	} while (0);
+
+	if (err < 0) {
+		ffmpeg_free_context(&cc);
+		avformat_close_input(&ic);
+		return err;
+	}
+
+	priv = xnew(struct ffmpeg_private, 1);
+	priv->codec_context = cc;
+	priv->input_context = ic;
+	priv->codec = codec;
+	priv->input = ffmpeg_input_create();
+	if (priv->input == NULL) {
+		ffmpeg_free_context(&cc);
+		avformat_close_input(&ic);
+		free(priv);
+		return -IP_ERROR_INTERNAL;
+	}
+	priv->input->stream_index = stream_index;
+	priv->output = ffmpeg_output_create();
+
+	cc->opaque = priv;
+
+	/* Prepare for resampling. */
+	swr = swr_alloc();
+	av_opt_set_int(swr, "in_channel_layout",  av_get_default_channel_layout(cc->channels), 0);
+	av_opt_set_int(swr, "out_channel_layout", av_get_default_channel_layout(cc->channels), 0);
+	av_opt_set_int(swr, "in_sample_rate",     cc->sample_rate, 0);
+	av_opt_set_int(swr, "out_sample_rate",    cc->sample_rate, 0);
+	av_opt_set_sample_fmt(swr, "in_sample_fmt",  cc->sample_fmt, 0);
+	priv->swr = swr;
+
+	ip_data->private = priv;
+	ip_data->sf = sf_rate(cc->sample_rate) | sf_channels(cc->channels);
+	switch (cc->sample_fmt) {
+	case AV_SAMPLE_FMT_U8:
+		ip_data->sf |= sf_bits(8) | sf_signed(0);
+		av_opt_set_sample_fmt(swr, "out_sample_fmt", AV_SAMPLE_FMT_U8,  0);
+		break;
+	case AV_SAMPLE_FMT_S32:
+		ip_data->sf |= sf_bits(32) | sf_signed(1);
+		av_opt_set_sample_fmt(swr, "out_sample_fmt", AV_SAMPLE_FMT_S32,  0);
+		break;
+	/* AV_SAMPLE_FMT_S16 */
+	default:
```

Comment:
I think you're missing the point of this code. We're checking the sample format of the audio the decoder gives us, `cc->sample_fmt`, and set an appropriate target format for resampling that can be handled by the rest of cmus. In fact, I assume that the code would be just fine if we simply replaced this whole `switch` by the `default` case.